### PR TITLE
Document stdlib

### PR DESCRIPTION
**Pull request by zimbatm**
I am trying to document the lib/ folder's functions because I always forget what argument I should pass and in what order. Before I do all of them, is this format acceptable ?
I don't really care about the syntax I just invented, just that it's consistent and used everywhere.

/cc @edolstra

### CODE REVIEW

**Comment by vcunat** [comment]
These are builtin functions, so we already document them in a different place. http://nixos.org/nix/manual/#ssec-builtins

IMO we should:
- have a common list of functions from `lib.*` and `builtins.*`, with their description, organized into sections (we have lib/*section*.nix);
- keep the descriptions next to the implementation and extract documentation from there. (We already have outdated docs of some builtins. I would hope this should improve the situation over the long term.)

**Comment by zimbatm** [comment]
What do you think of the type notation that I used ?

I can skip documentation of all the imported builtins for now.

**Comment by edolstra** [comment]
To be consistent, please follow the established format in `attrsets.nix`, e.g.
```
  /* Collect each attribute named `attr' from a list of attribute
     sets.  Sets that don't contain the named attribute are ignored.

     Example:
       catAttrs "a" [{a = 1;} {b = 0;} {a = 2;}]
       => [1 2]
  */
```

**Comment by zimbatm** [comment]
@edolstra better like that ?

I just finished documenting lib/attrsset and found two oddities:

* `matchAttrs { cpu = { bits = 64; }; } (mkSystem { })` throws "error: value is a built-in function while a set was expected" if the left matches is included in the right.

* deepSeqAttrs always returns `y` regardless of `x`. Is uses `deepSeqList` which has a similar behaviour. Since both are only used together I would be tempted to mark them as deprecated.

**Comment by domenkozar** [comment]
:cake:

**Comment by zimbatm** [comment]
@domenkozar I am running a bit out of steam doing this-one. I will just fix the remaining issues and declare this-one as gold. Let me know if you want a PR against 16.04 as well.

**Comment by zimbatm** [comment]
@domenkozar alright, I think this is good to go. There are still some FIXMEs but it's way better than before.

**Comment by domenkozar** [comment]
Looks good. @edolstra @shlevy

File: lib/attrsets.nix [hunk 1]
```diff
@@ -13,8 +13,14 @@ rec {
 
 
   /* Return an attribute from nested attribute sets.  For instance
```
Comment (shlevy) [inline lib/attrsets.nix]: "For instance" should be removed here now I think

File: lib/attrsets.nix [hunk 2]
```diff
@@ -248,11 +282,19 @@ rec {
...
+     { type = "derivation"; } counts as a derivation.
 
+     Example:
+       nixpkgs = import <nixpkgs> {}
+       isDerivation nixpkgs.ruby
+       => true
+       isDerivation "foobar"
+       => false
+  */
+  isDerivation = x: isAttrs x && x ? type && x.type == "derivation";
```
Comment (shlevy) [inline lib/attrsets.nix]: Side note: Should just have `isDerivation = x: (x.type or null) == "derivation"`

**Comment by shlevy** [comment]
This is pretty long :grinning: @domenkozar have you reviewed it all? If so :+1: from me.

**Comment by domenkozar** [comment]
I went through, but haven't tested all examples. I'll do that before the merge. (one day we should implement doctests)

File: lib/strings.nix [hunk 3]
```diff
@@ -119,21 +220,50 @@ rec {
...
+  /* Converts an ASCII string to upper-case.
+
+     Example:
+       toLower "home"
+       => "HOME"
+  */
   toUpper = replaceChars lowerChars upperChars;
 
+  /* Appends string context from another string.  This is an implementation
+     detail of nix.
```
Comment (edolstra) [inline lib/strings.nix]: `nix` -> `Nix`.

File: lib/strings.nix [hunk 4]
```diff
@@ -119,21 +220,50 @@ rec {
...
+
+     Strings in nix carry an invisible `context` which is a list of strings
```
Comment (edolstra) [inline lib/strings.nix]: No ` as a quote character please.

File: lib/attrsets.nix (hunk 2 above)
Comment (zimbatm) [inline lib/attrsets.nix]: You're right but I want to avoid introducing code changes in this PR

File: lib/attrsets.nix (hunk 1 above)
Comment (zimbatm) [inline lib/attrsets.nix]: Fixed in bc7d6deee450dc5297c707dc3ade1a4e0789c4f7

File: lib/strings.nix (hunk 4 above)
Comment (zimbatm) [inline lib/strings.nix]: Fixed in 4a3a4cb58e0fd7cc3a96918514c9eb70e23a8812

File: lib/strings.nix (hunk 3 above)
Comment (zimbatm) [inline lib/strings.nix]: Fixed in 4a3a4cb58e0fd7cc3a96918514c9eb70e23a8812

**Comment by zimbatm** [comment]
Most of the examples were generated using `nix-repl` then `:l ./lib` to load the context.

**Comment by zimbatm** [comment]
Let me know when it's good and I'll squash those fixup commits.

File: lib/strings.nix [hunk 5]
```diff
@@ -119,21 +220,50 @@ rec {
...
 
-  # Appends string context from another string.
+     Example:
+       pkgs = import <nixpkgs> { };
+       addContextFrom pkgs.coreutils "bar"
+       => "bar"
+  */
   addContextFrom = a: b: substring 0 0 a + b;
 
+  /* Cut a string with a separator and produces a list of strings which
```
Comment (edolstra) [inline lib/strings.nix]: This function is also really slow so should have a warning.

File: lib/strings.nix [hunk 6]
```diff
@@ -223,25 +401,56 @@ rec {
...
     if builtins.isInt may_be_int
     then may_be_int
     else throw "Could not convert ${str} to int.";
 
-  # Read a list of paths from `file', relative to the `rootPath'. Lines
-  # beginning with `#' are treated as comments and ignored. Whitespace
-  # is significant.
+  /* Read a list of paths from `file', relative to the `rootPath'. Lines
+     beginning with `#' are treated as comments and ignored. Whitespace
+     is significant.
```
Comment (edolstra) [inline lib/strings.nix]: This function sounds like a really bad idea performance-wise so I'm not sure it should be documented at all. Maybe it should be moved to the misc.nix ghetto.

File: lib/strings.nix (hunk 5 above)
Comment (zimbatm) [inline lib/strings.nix]: Fixed in ecae390163bf43f2ea20609efe11d8dca9cef59a

File: lib/strings.nix (hunk 6 above)
Comment (zimbatm) [inline lib/strings.nix]: I added a warning for now in 526efc622da51dd809a69943040e2c9b723235a3

**Comment by zimbatm** [comment]
Just squashed all the fixup commits, I think it's good to go now.

**Comment by zimbatm** [comment]
Ping, can I merge this before it gets stale ?

**Comment by domenkozar** [comment]
Give me another few days  :/

File: lib/lists.nix [hunk 7]
```diff
@@ -37,13 +52,22 @@ rec {
...
   foldl' = builtins.foldl' or foldl;
 
+  /* Map with index
+
+     FIXME(zimbatm): why does this start to count at 1?
 
-  # Map with index: `imap (i: v: "${v}-${toString i}") ["a" "b"] ==
-  # ["a-1" "b-2"]'. FIXME: why does this start to count at 1?
+     Example:
+       map (i: v: "${v}-${toString i}") ["a" "b"]
```
Comment (domenkozar) [inline lib/lists.nix]: shouldn't this be `imap`?

File: lib/lists.nix [hunk 8]
```diff
@@ -57,73 +81,141 @@ rec {
             else [ (f (n + 1) (elemAt list n)) ] ++ imap' (n + 1);
       in imap' 0;
 
+  /* Map and concatenate the result.
 
-  # Map and concatenate the result.
+     Example:
+       concatMap (x: x ++ ["z"]) ["a" "b"]
+       => [ "a" "z" "b" "z" ]
```
Comment (domenkozar) [inline lib/lists.nix]: nix-repl> concatMap (x: x ++ ["z"]) ["a" "b"]
error: value is a string while a list was expected, at (string):1:15

File: lib/lists.nix (hunk 7 above)
Comment (zimbatm) [inline lib/lists.nix]: good catch, fixing

File: lib/lists.nix [hunk 9]
```diff
@@ -57,73 +81,141 @@ rec {
...
+     spliced into the top-level lists.
 
-  # Flatten the argument into a single list; that is, nested lists are
-  # spliced into the top-level lists.  E.g., `flatten [1 [2 [3] 4] 5]
-  # == [1 2 3 4 5]' and `flatten 1 == [1]'.
+     Example:
+       flatten [1 [2 [3] 4] 5]
+       => [1 2 3 4 5]
+       flatten 1
+       => [1]'.
```
Comment (domenkozar) [inline lib/lists.nix]: typo at the end?

File: lib/lists.nix (hunk 7 above)
Comment (zimbatm) [inline lib/lists.nix]: Fixed in 40d417afc8fe9b0ef7369288d95b7b5f77fa0536

File: lib/lists.nix (hunk 8 above)
Comment (zimbatm) [inline lib/lists.nix]: Fixed in 79fe92df00cca9e68df41d8445801ce1bb59121a

File: lib/lists.nix (hunk 9 above)
Comment (zimbatm) [inline lib/lists.nix]: fixed in e1994a091e18366be14d98edc0affc598d568328

File: lib/strings.nix [hunk 10]
```diff
@@ -10,65 +10,147 @@ rec {
 
   inherit (builtins) stringLength substring head tail isString replaceStrings;
 
+  /* Concatenate a list of strings.
 
-  # Concatenate a list of strings.
+     Example:
+       concatString ["foo" "bar"]
```
Comment (domenkozar) [inline lib/strings.nix]: concatString -> concatStrings

File: lib/strings.nix [hunk 11]
```diff
@@ -10,65 +10,147 @@ rec {
...
+  */
   makeLibraryPath = makeSearchPath "lib";
 
-  # Construct a binary search path (such as $PATH) containing the
-  # binaries for a set of packages, e.g. "${pkg1}/bin:${pkg2}/bin:...".
+  /* Construct a binary search path (such as $PATH) containing the
+     binaries for a set of packages.
+
+     Example:
+       makeSearchPath ["/root" "/usr" "/usr/local"]
```
Comment (domenkozar) [inline lib/strings.nix]: makeSearchPath -> makeBinPath

File: lib/strings.nix (hunk 10 above)
Comment (zimbatm) [inline lib/strings.nix]: Fixed in 9eea475ae52268781161c76bf7ecd73768e19644

File: lib/strings.nix (hunk 11 above)
Comment (zimbatm) [inline lib/strings.nix]: Fixed in eb1d4b86501cce1bfec93f75f6856fd1e868c8f0

**Comment by domenkozar** [comment]
Cool, if you auto squash I'll merge.

**Comment by zimbatm** [comment]
@domenkozar thanks for reviewing, all squashed now

**Comment by domenkozar** [comment]
Thank you :cake:

**Comment by zimbatm** [comment]
:)

**Pull request merged**